lib/pymedphys/_dicom/connect/listen.py

```diff
@@ -0,0 +1,207 @@
+        self.association_directory = None
+
+        # The application entity
+        self.ae = None
+
+        logging.debug("Will store files received in: %s", self.storage_directory)
+
+    def start(self):
+        """Start the DICOM listener
+        """
+
+        # Initialise the Application Entity
+        self.ae = pynetdicom.AE(ae_title=self.ae_title)
+
+        # Add the supported presentation context
+        self.ae.add_supported_context(
+            pynetdicom.sop_class.VerificationSOPClass  # pylint: disable = no-member
+        )
+        for context in pynetdicom.StoragePresentationContexts:
+            self.ae.add_supported_context(context.abstract_syntax)
+
+        handlers = [
+            (pynetdicom.evt.EVT_C_STORE, self.on_c_store),
+            (pynetdicom.evt.EVT_C_ECHO, self.on_c_echo),
+            (pynetdicom.evt.EVT_ACCEPTED, self.on_association_accepted),
+            (pynetdicom.evt.EVT_RELEASED, self.on_association_released),
+        ]
+
+        # Start listening for incoming association requests
+        self.ae.start_server((self.host, self.port), evt_handlers=handlers, block=False)
+
+    def stop(self):
+        """Stop the DICOM listener
+        """
+
+        if self.ae:
+            self.ae.shutdown()
+
+    def on_c_echo(self, _):  # pylint: disable = no-self-use
+        """Respond to a C-ECHO service request.
+        """
+        logging.debug("C-ECHO!")
+        return 0x0000
+
+    def on_association_accepted(self, _):
+        self.association_directory = None
+
+    def on_association_released(self, _):
+        if self.on_released_callback:
+            self.on_released_callback(self.association_directory)
+
+    def on_c_store(self, event):
+
+        dataset = event.dataset
+
+        try:
+            mode_prefix = DICOM_SOP_CLASS_NAMES_MODE_PREFIXES[dataset.SOPClassUID.name]
+        except KeyError:
+            mode_prefix = "UN"
+
+        series_uid = dataset.SeriesInstanceUID
+        series_dir = self.storage_directory.joinpath(series_uid)
+        series_dir.mkdir(exist_ok=True)
+        self.association_directory = series_dir
+
+        filename = "{0!s}.{1!s}".format(mode_prefix, dataset.SOPInstanceUID)
+        filepath = series_dir.joinpath(filename)
+
+        status_ds = pydicom.Dataset()
+        status_ds.Status = 0x0000
+
+        if filepath.exists():
+
+            # If the file already exists, open it up and compare the hash of it's
+            # contents (converting to JSON string to perform hash)
+            existing_ds = pydicom.read_file(filepath)
+            existing_hash = hash(existing_ds.to_json())
+            incoming_hash = hash(dataset.to_json())
+
+            if not existing_hash == incoming_hash:
+                # If the hashes don't match, save conflicting incoming file in an
+                # "orphan" sub-directory.
+
+                # Just give the file a unique name in the orphan directory
+                filename = str(uuid.uuid4())
```

Comment:
Could this have the original filename with the random characters appended?

```suggestion
                filename = f"{filename}-{uuid.uuid4()}"
```

If recommendation above to include extension is undergone, would need:

```suggestion
                filename = pathlib.Path(
                    f"{filename.stem}-{uuid.uuid4()}"
                )
```